text/0000-floating-point-additional-precision.md

```diff
@@ -0,0 +1,141 @@
+calculations using more precision than the type itself, as long as they provide
+*at least* as much precision as the IEEE 754 standard requires.
+
+See the [prior art section](#prior-art) for precedent in several other
+languages and compilers.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+After this RFC, we could explain floating-point operations as follows:
+
+Floating-point operations in Rust have a guaranteed minimum accuracy, which
+specifies how far the result may differ from an infinitely accurate,
+mathematically exact answer. The implementation of Rust for any target platform
+must provide at least that much accuracy. In some cases, Rust can perform
+operations with higher accuracy than required, and doing so provides greater
+performance (such as by removing intermediate rounding steps). This will never
+make any floating-point operation *less* accurate, but it can make
+floating-point operations *more* accurate, making the result closer to the
+mathematically exact answer.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Currently, Rust's [specification for floating-point
+types](https://doc.rust-lang.org/reference/types/numeric.html#floating-point-types)
+states that:
+> The IEEE 754-2008 "binary32" and "binary64" floating-point types are f32 and f64, respectively.
+
+This RFC proposes updating that definition as follows:
+
+The `f32` and `f64` types represent the IEEE 754-2008 "binary32" and "binary64"
+floating-point types. Operations on those types must provide no less
+precision than the IEEE standard requires; such operations may provide *more*
+precision than the standard requires, such as by doing a series of operations
+with higher precision before storing a value of the desired precision.
+
+rustc may provide a codegen (`-C`) option to disable this behavior, such as `-C
+disable-extra-fp-precision`. Rust may also provide an attribute to disable this
+behavior from within code, such as `#[disable_extra_fp_precision]`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+If Rust already provided bit-for-bit identical floating-point computations
+across platforms, this change could potentially allow floating-point
+computations to differ by platform (though never below the standards-required
+accuracy). However, standards-compliant implementations of math functions on
+floating-point values may already vary slightly by platform, sufficiently so to
+produce different binary results. This proposal can never make results *less*
+accurate, it can only make results *more* accurate.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could provide a separate set of types and allow extra accuracy in their
+operations; however, this would create ABI differences between floating-point
+functions, and the longer, less-well-known types seem unlikely to see
```

Comment:
Not necessarily, these wrappers could be `repr(transparent)`.

Reply:
I mean this in the sense that changing from one to the other would be an incompatible API change in a crate.

Reply:
If the algorithm does not care about contraction, it might also not care about `NaN`s, so if it wants to accept a `NonNaN` type as well as the primitive `f{32, 64}` types, then it has to be generic, and if its generic, it would also accept a type wrapper lifting the assumption that contraction is not ok without breaking the API.

In other words, once one starts walking down the road of lifting assumptions about floating-point arithmetic, contraction is just one of the many many different assumptions that one might want to lift. Making it special does not solve the issue of these APIs having to be generic about these.

Reply:
I do not think we have anywhere near a smooth enough UX for working with wrappers around primitive arithmetic types for me to seriously considered them as a solution for licensing fast-math transformations. There's serious papercuts even when trying to generic over the existing primitive types (e.g., you can't use literals without wrapping them in ugly `T::from` calls), and we have even less machinery to address the *mixing* of different types that such wrappers would entail.

I also think it's quite questionable whether these should be properties of the *type*. It kind of fits "no infinities/nans/etc." but other things are fundamentally about particular operations and therefore may be OK in one code region but not in another code region even if the same data is being operated on.